pkg/appmanager/appManager.go

```diff
@@ -1145,6 +1146,158 @@ func (appMgr *Manager) syncIngresses(
+		ingress := obj.(*v1beta1.Ingress)
+		if ingress.Spec.Rules != nil {
+			if new != "" {
+				oldAnnotationMap := parseAppRootURLRewriteAnnotations(old)
+				newAnnotationMap := parseAppRootURLRewriteAnnotations(new)
+				var oldRuleNames string
+				var newRuleNames string
+				if isAppRoot {
+					oldRuleNames = generateMultiServiceAnnotationRuleNames(ingress, oldAnnotationMap, appRootRedirectRulePrefix)
+					oldRuleNames += "," + generateMultiServiceAnnotationRuleNames(ingress, oldAnnotationMap, appRootForwardRulePrefix)
+					newRuleNames = generateMultiServiceAnnotationRuleNames(ingress, newAnnotationMap, appRootRedirectRulePrefix)
+					newRuleNames += "," + generateMultiServiceAnnotationRuleNames(ingress, newAnnotationMap, appRootForwardRulePrefix)
+					ruleNames = computeAnnotationDiff(newRuleNames, oldRuleNames)
+				} else {
+					oldRuleNames = generateMultiServiceAnnotationRuleNames(ingress, oldAnnotationMap, urlRewriteRulePrefix)
+					newRuleNames = generateMultiServiceAnnotationRuleNames(ingress, newAnnotationMap, urlRewriteRulePrefix)
+					ruleNames = computeAnnotationDiff(newRuleNames, oldRuleNames)
+				}
+			} else {
+				annotationMap := parseAppRootURLRewriteAnnotations(old)
+				if isAppRoot {
+					ruleNames = generateMultiServiceAnnotationRuleNames(ingress, annotationMap, appRootRedirectRulePrefix)
+					ruleNames += "," + generateMultiServiceAnnotationRuleNames(ingress, annotationMap, appRootForwardRulePrefix)
+				} else {
+					ruleNames = generateMultiServiceAnnotationRuleNames(ingress, annotationMap, urlRewriteRulePrefix)
+				}
+			}
+		} else {
+			if isAppRoot {
+				annotationMap := parseAppRootURLRewriteAnnotations(old)
+				nameEnd := "single-service" + "-" + annotationMap["single"]
+				ruleNames = appRootRedirectRulePrefix + nameEnd
+				ruleNames += "," + appRootForwardRulePrefix + nameEnd
+			}
+		}
+	default:
+		log.Errorf("Unknown object type: %v", t)
+	}
+	return ruleNames
+}
+
+func processAnnotationDependencies(added, removed []ObjectDependency, obj interface{}) (bool, bool) {
+	var urlAdded ObjectDependency
+	var appAdded ObjectDependency
+	for i := range added {
+		if added[i].Kind == "URL-Rewrite-Annotation" {
+			urlAdded = added[i]
+		}
+		if added[i].Kind == "App-Root-Annotation" {
+			appAdded = added[i]
+		}
+	}
+
+	var urlRemoved ObjectDependency
+	var appRemoved ObjectDependency
+	for i := range removed {
+		if removed[i].Kind == "URL-Rewrite-Annotation" {
+			urlRemoved = removed[i]
+		}
+		if removed[i].Kind == "App-Root-Annotation" {
+			appRemoved = removed[i]
+		}
+	}
+
+	var urlUpdate bool
+	if urlAdded.Name != "" && urlRemoved.Name != "" {
+		urlUpdate = true
+		urlRemoved.Name = getRemovedAnnotationRuleNames(urlAdded.Name, urlRemoved.Name, false, obj)
+	} else if urlAdded.Name != "" && urlRemoved.Name == "" {
+		urlUpdate = true
+	} else if urlAdded.Name == "" && urlRemoved.Name != "" {
+		urlRemoved.Name = getRemovedAnnotationRuleNames("", urlRemoved.Name, false, obj)
```

Comment:
These urlRemoved and appRemoved aren't returned from the function; where are they actually used?